### DISCUSSION TITLE
Automatically add a timestamp to the log! macro

### DISCUSSION THREAD

**Discussion by wrm65** [leptos-rs/leptos#2144] [Ideas]
We all know how convenient (and necessary) it is to use the `console log` to figure out what our _well crafted_ and _it-should-work_ code is doing.

While its not a big deal, I also tend to find that including a timestamp (and the module name) to the log - an added bonus.

I have extended the `log!` macro to automatically prefix the logged message with the `timestamp info`. The timestamp format is `YYYY-MM-DD HH:MM:SS.nnn` with an added 3 decimal place microsecond. The timestamp is book-ended by a pair of squared brackets `[]` and the text is shown in the color - `green`.

**Here is an example of a timestamped `console_log` messaage:**

`[2023-12-30 01:03:43.814] ::PageOrientation::set_orientation::`

**and here is the extended macro, which I have called `tslog`.**

```
use leptos_dom::{logging::console_log};
use chrono::prelude::*;

#[macro_export]
macro_rules! tslog {
    ($($t:tt)*) => {
        (leptos_dom::logging::console_log(&format!("\x1B[32m[{}] \x1B[0m{}", 
                                          chrono::Local::now().format("%Y-%m-%d %H:%M:%S%.3f").to_string(), 
                                          format_args!($($t)*).to_string())))
    }
}
```
`\x1B[32m` is the escape sequence for setting the text color to blue, and
`\x1B[0m` reset the color (to black in this case) 

**and to use it,**

```
        let mut _text = format!("::{}::{}::",
                        self.logger, 
                        &function_name!().to_string());
        tslog!("{_text}");
```

**Comment by gbj** [comment]
Cool! The log macro is intended as a very thin wrapper over println (on the server) or console.log (in the browser), because println doesn't work in the browser. For any real logging purposes I'd recommend using the tracing or log  crates.

**Reply by wrm65** [reply]
Thanks for the recommendation!
